community-build/test/scala/dotty/communitybuild/CommunityBuildTest.scala

```diff
@@ -7,207 +7,223 @@ import org.junit.{Ignore, Test}
-    // Workaround for https://github.com/sbt/sbt/issues/4395
-    new File(sys.props("user.home") + "/.sbt/1.0/plugins").mkdirs()
-    val pluginFilePath = communitybuildDir.resolve("sbt-dotty-sbt").toAbsolutePath().toString()
+lazy val compilerVersion: String =
+  val file = communitybuildDir.resolve("dotty-bootstrapped.version")
+  new String(Files.readAllBytes(file), UTF_8)
 
-    // Run the sbt command with the compiler version and sbt plugin set in the build
-    val arguments = {
-      val sbtProps = Option(System.getProperty("sbt.ivy.home")) match {
-        case Some(ivyHome) =>
-          Seq(s"-Dsbt.ivy.home=$ivyHome")
-        case _ =>
-          Seq()
-      }
-      extraSbtArgs ++ sbtProps ++ Seq(
-        "-sbt-version", "1.2.7",
-        s"--addPluginSbtFile=$pluginFilePath",
-        s";clean ;set updateOptions in Global ~= (_.withLatestSnapshots(false)) ;++$compilerVersion! $testCommand"
-      )
-    }
+lazy val sbtPluginFilePath: String =
+  // Workaround for https://github.com/sbt/sbt/issues/4395
+  new File(sys.props("user.home") + "/.sbt/1.0/plugins").mkdirs()
+  communitybuildDir.resolve("sbt-dotty-sbt").toAbsolutePath().toString()
 
-    test(project, "sbt", arguments)
-  }
+def log(msg: String) = println(Console.GREEN + msg + Console.RESET)
 
-  def testMill(project: String, testCommand: String, extraMillArgs: Seq[String] = Nil) =
-    test(project, "./mill", extraMillArgs :+ testCommand)
+/** Executes shell command, returns false in case of error. */
+def exec(projectDir: Path, binary: String, arguments: String*): Int =
+  val command = binary +: arguments
+  log(command.mkString(" "))
+  val builder = new ProcessBuilder(command: _*).directory(projectDir.toFile).inheritIO()
+  val process = builder.start()
+  val exitCode = process.waitFor()
+  exitCode
 
-  /** Build the given project with the published local compiler and sbt plugin.
-   *
-   *  This test reads the compiler version from community-build/dotty-bootstrapped.version
-   *  and expects community-build/sbt-dotty-sbt to set the compiler plugin.
-   *
-   *  @param project        The project name, should be a git submodule in community-build/
-   *  @param command        The sbt command used to test the project
-   *  @param updateCommand  The sbt command used to update the project
-   *  @param extraSbtArgs   Extra arguments to pass to sbt
-   */
-  def test(project: String, command: String, arguments: Seq[String]): Unit = {
-    def log(msg: String) = println(Console.GREEN + msg + Console.RESET)
 
-    log(s"Building $project with dotty-bootstrapped $compilerVersion...")
+sealed trait CommunityProject
+  private var published = false
 
-    val projectDir = communitybuildDir.resolve("community-projects").resolve(project)
+  val project: String
+  val updateCommand: String
+  val testCommand: String
+  val publishCommand: String
+  val dependencies: List[CommunityProject]
```

Comment:
If the versions of depended libs do not match the dep specs of the depending lib, publishing will not guarantee that the published artifacts will be used.

It might not be an issue now, we may address it later as the community build evolves.

Reply:
Good point. I also think that if it doesn't bother us right now, it is best to postpone it till we have actual use cases.